--- swell/swell-ini.cpp	Wed Jun 13 13:46:01 2012
+++ swell/swell-ini.cpp	Fri Jul 13 15:06:21 2012
@@ -324,6 +324,24 @@
   return TRUE;
 }
 
+char* lstrcpyn_trimmed(char* dest, const char* src, int len)
+{
+	if (len<1)
+		return dest;
+	// Mimic Win32 behavior of stripping quotes and whitespace
+	while(*src==' ' || *src=='\t') ++src; // Strip beginning whitespace
+	const char* end = src+strlen(src)-1;
+	while(end >= src && (*end==' ' || *end=='\t')) --end; // Strip end whitespace
+	if (end > src && ((*src=='\"' && *end=='\"') || (*src=='\'' && *end=='\'')))
+	{	// Strip initial set of "" or ''
+		++src; --end;
+	}
+	int newlen = end-src+2;
+	if (newlen > len)
+		newlen = len;
+	return lstrcpyn(dest, src, newlen);
+}	
+
 DWORD GetPrivateProfileSection(const char *appname, char *strout, DWORD strout_len, const char *fn)
 {
   WDL_MutexLock lock(&m_mutex);
@@ -355,10 +373,11 @@
         
         WRSTR(kv)
         WRSTR("=")
-        WRSTR(val)
-        
 #undef WRSTR
 
+	    lstrcpyn_trimmed(strout+szOut, val, strout_len - szOut - 2);
+        szOut += strlen(strout+szOut);
+
         l=1;
         if (l > strout_len - szOut - 1) l = strout_len - 1 - szOut;
         if (l>0) { memset(strout+szOut,0,l); szOut+=l; }
@@ -431,13 +450,13 @@
       const char *val = cursec->Get(keyname);
       if (val)
       {
-        lstrcpyn_safe(ret,val,retsize);
+        lstrcpyn_trimmed(ret,val,retsize);
         return strlen(ret);
       }
     }
   }
 //  printf("def %s %s %s %s\n",appname,keyname,def,fn);
-  lstrcpyn_safe(ret,def?def:"",retsize);
+  lstrcpyn_trimmed(ret,def?def:"",retsize);
   return strlen(ret);
 }
 
