--- swell/swell-ini.cpp	Mon Jan 19 16:26:36 1970
+++ swell/swell-ini.cpp	Mon Jan 19 16:26:36 1970
@@ -168,7 +168,15 @@
         {
           *t++=0;
           if (*p) 
+          {
+            // Mimic Win32 behavior of stripping quotes and whitespace
+            while(*t==' ' || *t=='\t') ++t;
+            char* e = t+strlen(t)-1;
+            while(e >= t && (*e==' ' || *e=='\t')) --e;
+            if (e > t && ((*t=='\"' && *e=='\"') || (*t=='\'' && *e=='\''))) { ++t; --e; }
+            e[1] = 0;
             cursec->AddUnsorted(p,strdup(t));
+          }
         }
       }
     }
@@ -314,7 +322,7 @@
   int szOut=0;
   WDL_StringKeyedArray<char *> *cursec = ctx ? ctx->m_sections.Get(appname) : NULL;
 
-  if (ctx) 
+  if (ctx && cursec)
   {
     int x;
     for(x=0;x<cursec->GetSize();x++)
--- wingui/virtwnd-controls.h	Mon Jan 19 16:26:36 1970
+++ wingui/virtwnd-controls.h	Mon Jan 19 16:26:36 1970
@@ -115,8 +115,9 @@
     bool GetForceText() { return m_forcetext; }
 
     void SetFont(LICE_IFont *font, LICE_IFont *vfont=NULL) { m_textfont=font; m_textfontv=vfont; }
+    LICE_IFont *GetFont(bool vfont=false) { return vfont?m_textfontv:m_textfont; }
 
-  private:
+  protected:

     void DoSendCommand(int xpos, int ypos);
 